feat: release keyPress automatically

ym opened the pull request.

ym committed:
feat: release keyPress automatically

ym committed:
send keepalive when pressing the key

ym committed:
remove logging

ym committed:
clean up logging

ym committed:
chore: use unreliable channel to send keepalive events

ym committed:
chore: use ordered unreliable channel for pointer events

ym committed:
chore: adjust auto release key interval

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -173,6 +178,64 @@ func (u *UsbGadget) SetOnKeysDownChange(f func(state KeysDownState)) {
+	u.kbdAutoReleaseLock.Lock()
+	defer u.kbdAutoReleaseLock.Unlock()
+
+	u.log.Trace().Msg("delaying auto-release")
```
_nit_ we should probably emit the trace _before_ trying to grab the lock... would be easier to see deadlocks if they ever happen.

IDisposable commented on ui/src/hooks/useKeyboard.ts:
```diff
@@ -111,12 +104,61 @@ export default function useKeyboard() {
     }
   };
 
+  const KEEPALIVE_INTERVAL = 75; // 200ms interval
```
How is `75` related to `200ms`?

ym committed:
chore: update logging for kbdAutoReleaseLock

ym committed:
chore: update comment for KEEPALIVE_INTERVAL

adamshiervani commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -22,6 +22,11 @@ var keyboardConfig = gadgetConfigItem{
 
+// macOS default: 15 * 15 = 225ms https://discussions.apple.com/thread/1316947?sortBy=rank
+// Linux default: 250ms https://man.archlinux.org/man/kbdrate.8.en
+// Windows default: 1ms `HKEY_CURRENT_USER\Control Panel\Accessibility\Keyboard Response\AutoRepeatDelay`
```
nit: 1s on windows

adamshiervani commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -398,5 +464,26 @@ func (u *UsbGadget) KeypressReport(key byte, press bool) (KeysDownState, error)
+		{
+			u.log.Trace().Msg("acquiring kbdAutoReleaseLock to update last key")
+			u.kbdAutoReleaseLock.Lock()
+			u.kbdAutoReleaseLastKey = key
```
Not needed anymore as you reset the entire keyboard state on autorelease

IDisposable replied:
> Not needed anymore as you reset the entire keyboard state on autorelease

Seem it is now _correctly_ is [now using this value](https://github.com/jetkvm/kvm/pull/796/files/cbf1c7fba68dc9c4a0a083537444bc85d3a28020..12210fac96ef7b9ddc4d0bdf4c7e476e7455505a#diff-ca2c65f26d876c5c490651ad24cf7b6cbc9a996b1d969d2d9658c7d439351a8cR226-R237)

adamshiervani commented on internal/usbgadget/usbgadget.go:
```diff
@@ -68,6 +68,10 @@ type UsbGadget struct {
 
+	kbdAutoReleaseLock    sync.Mutex
+	kbdAutoReleaseTimer   *time.Timer
+	kbdAutoReleaseLastKey byte
```
Same

adamshiervani commented on ui/src/hooks/useKeyboard.ts:
```diff
@@ -111,12 +104,61 @@ export default function useKeyboard() {
+    if (keepAliveTimerRef.current) {
+      clearInterval(keepAliveTimerRef.current);
+    }
+
```
Call `sendKeypressKeepAliveHidRpc()` immediately and don't wait for the first `setInterval`.

adamshiervani commented on webrtc.go:
```diff
@@ -157,40 +205,12 @@ func newSession(config SessionConfig) (*Session, error) {
-				}
-			})
+			d.OnMessage(getOnHidMessageHandler(session, scopedLogger, "hidrpc"))
+		// we won't send anything over the unreliable channels
```
Could you clarify this a bit more?

ym committed:
fix: should cancelAutorelease when pressed is true

ym committed:
fix: handshake won't happen if webrtc reconnects

ym committed:
chore: add trace log for writeWithTimeout

ym committed:
chore: add timeout for KeypressReport

ym committed:
chore: use the proper key to send release command

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -215,9 +219,12 @@ func (u *UsbGadget) DelayAutoRelease() {
+	u.log.Trace().Msg("performing autoRelease")
 	u.kbdAutoReleaseLock.Lock()
-	defer u.kbdAutoReleaseLock.Unlock()
+	defer unlockWithLog(&u.kbdAutoReleaseLock, u.log, "autoRelease unlocked")
```
s/unlocked/released

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -226,7 +233,8 @@ func (u *UsbGadget) performAutoRelease(key byte) {
 	// we just reset the keyboard state to 0 no matter what
-	_, err := u.keypressReport(0, false, false)
+	u.log.Trace().Uint8("key", key).Msg("auto-releasing keyboard key")
+	_, err := u.keypressReport(key, false, false)
```
👍

IDisposable commented on internal/usbgadget/hid_keyboard.go:
Should this be `.Bytes("keys", keys)` so we get the more readable logging?

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -481,6 +503,31 @@ func (u *UsbGadget) keypressReport(key byte, press bool, autoRelease bool) (Keys
+	u.keyboardLock.Lock()
+	defer u.keyboardLock.Unlock()
+
+	r := make(chan keypressReportResult)
```
Seems like a lot of setup and tear-down to execute every single keypress... can we preconstruct the channel and go routine during setup and just use it here?

IDisposable commented on internal/usbgadget/utils.go:
```diff
@@ -164,3 +171,8 @@ func (u *UsbGadget) resetLogSuppressionCounter(counterName string) {
 }
+
+func unlockWithLog(lock *sync.Mutex, logger *zerolog.Logger, msg string, args ...any) {
+	logger.Trace().Msgf(msg, args...)
+	lock.Unlock()
```
If this throws during the trace (unlikely?) then we would not release the lock. I suggest the normal pattern
```golang
	defer lock.Unlock()
    logger.Trace().Msgf(msg, args...)
```

IDisposable commented on ui/src/hooks/useHidRpc.ts:
```diff
@@ -167,10 +167,24 @@ export function useHidRpc(onHidRpcMessage?: (payload: RpcMessage) => void) {
     };
 
+    const openHandler = () => {
+      console.warn("HID RPC channel opened");
```
Warn? Seems like this is more Trace/Info not an error state...

IDisposable commented on ui/src/hooks/useHidRpc.ts:
```diff
@@ -167,10 +167,24 @@ export function useHidRpc(onHidRpcMessage?: (payload: RpcMessage) => void) {
+    };
+
+    const closeHandler = () => {
+      console.warn("HID RPC channel closed");
```
Warn? Seems like this is more Trace/Info not an error state...

adamshiervani committed:
refactor: simplify HID RPC keyboard input handling and improve key state management

- Updated `handleHidRPCKeyboardInput` to return errors directly instead of keys down state.
- Refactored `rpcKeyboardReport` and `rpcKeypressReport` to return errors instead of states.
- Introduced a queue for managing key down state updates in the `Session` struct to prevent input handling stalls.
- Adjusted the `UpdateKeysDown` method to handle state changes more efficiently.
- Removed unnecessary logging and commented-out code for clarity.

adamshiervani committed:
refactor: enhance keyboard auto-release functionality and key state management

adamshiervani committed:
fix: correct Windows default auto-repeat delay comment from 1ms to 1s

adamshiervani committed:
refactor: send keypress as early as possible

adamshiervani committed:
refactor: replace console.warn with console.info for HID RPC channel events

adamshiervani commented:
I've now fixed all the issues that I found while testing. There were a bunch of places where we got stuck, mainly caused by bad network or overloaded hidWrites.

I've just pushed all the fixes, and mark this to **Ready for Review**. In addition to a normal review, I suggest checking the code out and following the hidRPC code path - good to understand, but also try to find issues.

My testing was done by creating different combinations of `handleKeyPress(key, press)` in JS: 
```js
// Prints abc...xyz 3 times
for (let i = 0; i < 3; i++) {
  for (let i = 0; i < 26; i++) {
    handleKeyPress(keys[`Key${String.fromCharCode(65 + i)}`], true);
    await new Promise(resolve => setTimeout(resolve, 10));
    handleKeyPress(keys[`Key${String.fromCharCode(65 + i)}`], false);
    await new Promise(resolve => setTimeout(resolve, 10));
  }
}
```

Other than that, just smashing your keyboard on a bad network can be a good way to ensure the auto-release is doing what it should.


My testing was on Ubuntu 22.04, and I used [Network Link Conditioner](https://nshipster.com/network-link-conditioner/) to simulate both packet loss and delay.

@IDisposable @ym

adamshiervani committed:
refactor: remove unused NewKeypressKeepAliveMessage function from HID RPC

ym commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -22,6 +26,11 @@ var keyboardConfig = gadgetConfigItem{
+// macOS default: 15 * 15 = 225ms https://discussions.apple.com/thread/1316947?sortBy=rank
+// Linux default: 250ms https://man.archlinux.org/man/kbdrate.8.en
+// Windows default: 1s `HKEY_CURRENT_USER\Control Panel\Accessibility\Keyboard Response\AutoRepeatDelay`
+const autoReleaseKeyboardInterval = time.Millisecond * 100
```
Should we change it back around `225ms` ?

adamshiervani committed:
fix: handle error in key release process and log warnings

adamshiervani committed:
fix: log warning on keypress report failure

adamshiervani committed:
fix: update auto-release keyboard interval to 225

adamshiervani committed:
refactor: enhance keep-alive handling and jitter compensation in HID RPC

- Implemented staleness guard to ignore outdated keep-alive packets.
- Added jitter compensation logic to adjust timer extensions based on packet arrival times.
- Introduced new methods for managing keep-alive state and reset functionality in the Session struct.
- Updated auto-release delay mechanism to use dynamic durations based on keep-alive timing.
- Adjusted keep-alive interval in the UI to improve responsiveness.

ym committed:
gofmt

ym committed:
clean up code

ym committed:
chore: use dynamic duration for scheduleAutoRelease

IDisposable commented on hidrpc.go:
```diff
@@ -85,27 +91,88 @@ func onHidMessage(data []byte, session *Session) {
+				// This is likely a retransmit stall or ordering delay.
+				// We reject the tick entirely and DO NOT extend,
+				// so the auto-release still fires on time.
+				validTick = false
```
this can just be `return nil` since we're going to bail out at line 143 anyway (and thus that block and the flag can be deleted)

adamshiervani replied:
I totally understand your comment, but I'd like to keep it extremely verbose, as this part can be difficult to grasp, and for me that's just a tad easier to follow.

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -145,32 +149,105 @@ func (u *UsbGadget) GetKeysDownState() KeysDownState {
 
-		u.log.Trace().Interface("old", u.keysDownState).Interface("new", state).Msg("keysDownState updated")
-		u.keysDownState = state
+	if u.kbdAutoReleaseTimers == nil {
+		return
```
I can't see any path where this would be nil (the entire collection, not the individual timer). I don't think this check is needed.

adamshiervani replied:
Neither can I

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -145,32 +149,105 @@ func (u *UsbGadget) GetKeysDownState() KeysDownState {
-func (u *UsbGadget) SetOnKeysDownChange(f func(state KeysDownState)) {
-	u.onKeysDownChange = &f
+func (u *UsbGadget) performAutoRelease(key byte) {
+	u.kbdAutoReleaseLock.Lock()
```
This block could be better structured as a typical defer block if we surround with a scope block, thus:

```golang
// lock scope
{ 
    u.kbdAutoReleaseLock.Lock()
	defer u.kbdAutoReleaseLock.Unlock()

	if u.kbdAutoReleaseTimers[key] == nil {
		u.log.Warn().Uint8("key", key).Msg("autoRelease timer not found")
		return
	}

	u.kbdAutoReleaseTimers[key].Stop()
	u.kbdAutoReleaseTimers[key] = nil
	delete(u.kbdAutoReleaseTimers, key
} // end lock scope
```

adamshiervani replied:
It was something along those lines in a previous commit, but it was causing some issues. That’s when I realized defer isn’t block-scoped, only function-scoped.

IDisposable replied:
Oh, I was misinformed, so sorry... I thought it was block scoped in go. Nice to learn

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -145,32 +149,105 @@ func (u *UsbGadget) GetKeysDownState() KeysDownState {
+	u.kbdAutoReleaseLock.Unlock()
+
+	// Skip if already released
+	state := u.GetKeysDownState()
```
Should this have been done inside the lock scope?

adamshiervani replied:
Its lock is inside the call.

IDisposable commented on internal/usbgadget/usbgadget.go:
```diff
@@ -149,3 +156,35 @@ func newUsbGadget(name string, configMap map[string]gadgetConfigItem, enabledDev
+	u.kbdAutoReleaseLock.Lock()
+	for _, timer := range u.kbdAutoReleaseTimers {
+		timer.Stop()
+	}
```
In other paths, we explicitly `delete` the timers as we go... doing this way will not leak anything?

adamshiervani replied:
Checking for null. Thanks!

IDisposable commented on ui/src/components/WebRTCVideo.tsx:
```diff
@@ -237,6 +237,7 @@ export default function WebRTCVideo() {
   const keyDownHandler = useCallback(
     (e: KeyboardEvent) => {
       e.preventDefault();
+      if (e.repeat) return;
```
This means that we'll not reset the auto-key-release timers on the device while we _KNOW_ that the key is still being held down. That seems incorrect, since the only way we will be getting these `e.repeat == true` client-side events is if the key is still down. 

Wouldn't it be better to allow this to still be communicated in some way to the device so the auto-release doesn't fire?

adamshiervani replied:
We don’t need repeats - those only add complexity - just much easier to reason with fewer events. The backend already maintains the key state using KeypressKeepAliveReport.

For devices that support the keypress hidRPC, each key-down starts a setInterval that sends a KeypressKeepAliveReport every 50 ms. The corresponding key-up clears that interval and stops sending keep alives.

Legacy devices (no hidRPC) run in stateless mode: each JSON-RPC KeyboardReport fully specifies which keys are down, and the device holds that state until a new report arrives - no repeats or keep-alives required.

IDisposable commented on ui/src/hooks/useHidRpc.ts:
```diff
@@ -6,45 +6,101 @@ import {
+    return (
+      rpcHidUnreliableChannel?.readyState === "open" && rpcHidProtocolVersion !== null
+    );
+  }, [rpcHidUnreliableChannel, rpcHidProtocolVersion]);
```
Would this be a better dependency as `[rpcHidUnreliableChannel?.ready, rpcHidProtocolVersion]` (which is what VSCode would have inserted)?

IDisposable commented on ui/src/hooks/useHidRpc.ts:
```diff
@@ -6,45 +6,101 @@ import {
+      rpcHidUnreliableNonOrderedChannel?.readyState === "open" &&
+      rpcHidProtocolVersion !== null
+    );
+  }, [rpcHidUnreliableNonOrderedChannel, rpcHidProtocolVersion]);
```
same question as line 42

IDisposable commented on ui/src/hooks/useHidRpc.ts:
```diff
@@ -68,29 +126,36 @@ export function useHidRpc(onHidRpcMessage?: (payload: RpcMessage) => void) {
+        // we assume that the UI is always using the latest version of the HID RPC protocol
+        // so we can't support this
+        // TODO: use capabilities to determine rather than version number
+        console.error("Server is using a newer HID RPC version than the client", message);
```
We could trigger a forced reload of the client-side page here with `window.location.reload();` but if that results in the client still being behind the server... something is wonky.

adamshiervani replied:
Let's just keep it as is and think it through properly when we tackle the new/old UI and new/old backend task.

IDisposable commented on ui/src/hooks/useKeyboard.ts:
```diff
@@ -111,12 +114,60 @@ export default function useKeyboard() {
+  const scheduleKeepAlive = useCallback(() => {
+    // Clear existing timer if it exists
+    if (keepAliveTimerRef.current) {
+      clearInterval(keepAliveTimerRef.current);
```
should we also `delete keepAliveTimerRef.current;` to ensure there's not a reference cycle?

adamshiervani replied:
Calling `cancelKeepAlive()` instead.

IDisposable commented on ui/src/hooks/useKeyboard.ts:
```diff
@@ -111,12 +114,60 @@ export default function useKeyboard() {
+  }, [sendKeypressKeepAliveHidRpc]);
+
+  // resetKeyboardState is used to reset the keyboard state to no keys pressed and no modifiers.
+  // This is useful for macros and when the browser loses focus to ensure that the keyboard state
```
`for macros, in case of client-side rollover,  and when the browser loses focus`

IDisposable commented on webrtc.go:
```diff
@@ -78,16 +97,85 @@ func (s *Session) initQueues() {
+func (s *Session) initKeysDownStateQueue() {
+	// serialise outbound key state reports so unreliable links can't stall input handling
+	s.keysDownStateQueue = make(chan usbgadget.KeysDownState, keysDownStateQueueSize)
+	go s.handleKeysDownStateQueue()
```
Do we need a cancel flag to exit the queue's go-routine loop?

adamshiervani replied:
No need up here, BUT I realized that we weren't closing the channel on connection close, which I added.

IDisposable commented on webrtc.go:
```diff
@@ -78,16 +97,85 @@ func (s *Session) initQueues() {
+	}
+}
+
+const keysDownStateQueueSize = 256
```
This seems _really deep_ :)

adamshiervani replied:
😂

IDisposable commented on webrtc.go:
```diff
@@ -157,40 +246,12 @@ func newSession(config SessionConfig) (*Session, error) {
-					return
-				}
-			})
+			d.OnMessage(getOnHidMessageHandler(session, scopedLogger, "hidrpc"))
```
Since the label is part of the RPC message, would it be better to use shorter strings, especially for the newly introduced ones (since we don't have legacy support concerns), perhaps "hidr", "hiduo", and "hidun" or even "hr", "ho", and "hn" since the label is really meaningless outside this code.

adamshiervani replied:
It's also used for logs/debugging, and in general I really prefer senseful verbosity over terseness.

IDisposable reviewed (changes requested):
There's a couple suggestions, and a couple questions... feel free to disregard.

adamshiervani committed:
Use harcoded timer reset value for now

adamshiervani committed:
fix: prevent nil pointer dereference when stopping timers in Close method

adamshiervani committed:
refactor: remove nil check for kbdAutoReleaseTimers in DelayAutoReleaseWithDuration

adamshiervani committed:
refactor: optimize dependencies in useHidRpc hooks

adamshiervani committed:
refactor: streamline keep-alive timer management in useKeyboard hook

adamshiervani committed:
refactor: clarify comments in useKeyboard hook for resetKeyboardState function

adamshiervani committed:
refactor: reduce keysDownStateQueueSize

adamshiervani committed:
refactor: close and reset keysDownStateQueue in newSession function

ym committed:
chore: resolve conflicts

ym committed:
resolve conflicts
